peps/pep-0735.rst

```diff
@@ -0,0 +1,490 @@
+The ``packages`` table contains an arbitrary number of user-defined keys, each of
+which has, as its value, a list of requirements specifiers (defined below).
+These keys must match the following regular expression:
+``[a-z0-9][a-z0-9-]*[a-z0-9]``. Meaning that they must be all lower-case
+alphanumerics, with ``-`` allowed only in the middle, and at least
+two characters long. These requirements are chosen so that the normalization
+rules used for PyPI package names are unnecessary as the names are already
+normalized.
+
+Requirements specifiers will use a definition based on :pep:`508`. This PEP
+also proposes extending the syntax to define a string format, a
+"PEP 735 Dependency". These are defined as strings in one of the following
+formats:
+
+* a PEP 508 specification, e.g., ``numpy>1``
+* a comma-delimited list of Dependency Group names, enclosed in square
+  brackets, e.g., ``[test,docs]`` to refer to other Dependency Groups named
+  ``test`` and ``docs`` as specified in ``[requirements.packages]``
+* a single dot, ``.``, which refers to the current project as a package
+  (similar to ``pip install .``)
+* a single dot, followed by square brackets enclosing a list of extra names,
+  which refers to the current project as a package including some extras, e.g.,
+  ``.[mysql]`` to refer to the current package with its ``mysql`` extra
+
+Each dependency group maps a name to a list of PEP 735 Dependencies.
+
+Any additional keys in a requirement specifier object are reserved for future
+use.
+
+The following is an example of a partial ``pyproject.toml`` which uses this to
+define four dependency groups: ``test``, ``docs``, ``typing``, and
+``typing-test``:
+
+.. code:: toml
+
+    [requirements.packages]
+    test = ["pytest", "coverage", "."]
+    docs = ["sphinx", "sphinx-rtd-theme"]
+    typing = ["mypy", "types-requests", ".[types]"]
+    typing-test = ["[typing,test]", "useful-types"]
+
+    [project.optional-dependencies]
+    types = ["typing-extensions"]
+
+Note how ``test`` and ``typing`` are able to refer to the current package
+while ``docs`` does not. This reflects the ability of Dependency Groups to be used
+in the same manner as extras, adding to dependencies, or completely
+independently. ``typing-test`` is defined as a union of two existing groups,
+plus an additional package. ``typing`` includes an extra, ``types``.
+
+Implementation Requirements
+---------------------------
+
+Build backends MUST NOT include dependency group data in built distributions.
+
+When installing the current package (``.``) from a dependency group, tools
+SHOULD prefer editable installs over non-editable installs. They MAY provide
+users with options to configure this behavior.
+
+Use of Dependency Groups
+------------------------
+
+Tools which support Dependency Groups are expected to provide new options and
+interfaces to allow users to install from Dependency Groups. Implementations
+may wish to treat them similarly to their current treatments of
```

Comment:
```suggestion
may wish to treat them similarly to their current treatment of
```
Though this sentence feels very vague. If I was a tool author, I'm not sure what I would do with it.

Reply:
Yeah, let's strike the whole sentence.

I meant this based on my experience with tox, which allows either of these:
```ini
deps = -r test-requirements.txt
commands = pytest
```
or
```ini
extras = dev
commands = pytest
```

and all I wanted to say was that this PEP doesn't say _anything_ about how tox (or any other tool like it) should handle specification of dependency groups.